File: src/services/hex.ts
import { Hex, Loc } from "../types/hex";
import { qMax, hexSize } from "../data/constants";

function from(q: number, r: number, s: number): Hex {
  if (q + r + s !== 0) {
    throw new Error("not a hex");
  }

  return { q, r, s };
}

function add(a: Hex, b: Hex): Hex {
  return {
    q: a.q + b.q,
    r: a.r + b.r,
    s: a.s + b.s,
  };
}

function sub(a: Hex, b: Hex): Hex {
  return {
    q: a.q - b.q,
    r: a.r - b.r,
    s: a.s - b.s,
  };
}

function eq(a: Hex, b: Hex) {
  return a.q === b.q && a.r === b.r && a.s === b.s;
}

function distance(a: Hex, b: Hex): number {
  const hex = sub(a, b);

  return (Math.abs(hex.q) + Math.abs(hex.r) + Math.abs(hex.s)) / 2;
}

function toPixel(h: Hex): Loc {
  const x: number =
    (Math.sqrt(3.0) * h.q + (Math.sqrt(3.0) / 2.0) * h.r) * hexSize;
  const y: number = (3.0 / 2.0) * h.r * hexSize;

  return { x, y };
}

function polygonCorners(h: Hex): Loc[] {
  const corners: Loc[] = [];
  const center = toPixel(h);
  for (let i = 0; i < 6; i++) {
    const angle = (2.0 * Math.PI * (i + 0.5)) / 6.0;
    const x = Math.cos(angle) * hexSize + center.x;
    const y = Math.sin(angle) * hexSize + center.y;

    corners.push({ x, y });
  }

  return corners;
}

export const hexDirections = [
  from(1, 0, -1), // E
  from(1, -1, 0), // NE
  from(0, -1, 1), // NW
  from(-1, 0, 1), // W
  from(-1, 1, 0), // SW
  from(0, 1, -1), // SE
];

export const hexOrigin = from(0, 0, 0);

function toStr(h: Hex): string {
  return `${h.q},${h.r},${h.s}`;
}

export const hexMap: Hex[] = [];
for (let q = -qMax; q <= qMax; q++) {
  let r1 = Math.max(-qMax, -q - qMax);
  let r2 = Math.min(qMax, -q + qMax);
  for (let r = r1; r <= r2; r++) {
    hexMap.push(from(q, r, -q - r));
  }
}

function isMapCorner(h: Hex) {
  if (h.q === 0 || h.r === 0 || h.s === 0) {
    if (
      Math.abs(h.q) === qMax ||
      Math.abs(h.r) === qMax ||
      Math.abs(h.s) === qMax
    ) {
      return true;
    }
  }

  return false;
}

export const HexService = {
  from,
  add,
  sub,
  eq,
  distance,
  toPixel,
  polygonCorners,
  toStr,
  isMapCorner,
};

File: lib/hex.dart
import 'dart:math';
import 'package:flame/game.dart';
import 'package:flutter/foundation.dart' show immutable;

const oddqDirectionDiff = [
  // even cols
  [
    [1, 0],
    [1, -1],
    [0, -1],
    [-1, -1],
    [-1, 0],
    [0, 1],
  ],
  // odd cols
  [
    [1, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
    [-1, 1],
    [0, 1],
  ],
];

// odd‑q (flat‑top) vertical layout from redblob
@immutable
class Hex {
  static const double size = 60;
  static Hex zero = const Hex(0, 0);
  static Hex invalid = const Hex(-1, -1);

  final int x;
  final int y;

  const Hex(this.x, this.y);

  Hex operator +(Hex that) {
    return Hex(x + that.x, y + that.y);
  }

  Hex operator -(Hex that) {
    return Hex(x - that.x, y - that.y);
  }

  int distance(Hex that) {
    final int q = x - that.x;
    final int r =
        y - (x - (x & 1)) ~/ 2 - (that.y - (that.x - (that.x & 1)) ~/ 2);
    final int s = q + r;

    return (q.abs() + r.abs() + s.abs()) ~/ 2;
  }

  Vector2 toPixel() {
    final dx = (3 / 2) * x * size;
    final dy = sqrt(3) * (y + 0.5 * (x & 1)) * size;

    return Vector2(dx, dy);
  }

  Hex neighbour(int direction) {
    final parity = x & 1;
    final diff = oddqDirectionDiff[parity][direction];
    return Hex(x + diff[0], y + diff[1]);
  }

  List<Hex> getNeighbours() {
    return List.generate(6, (i) => neighbour(i), growable: false);
  }

  List<Vector2> polygonCorners([double size = Hex.size]) {
    final List<Vector2> corners = [];
    final center = toPixel();
    for (int i = 0; i < 6; i++) {
      final angle = (2.0 * pi * i) / 6.0;
      final x = cos(angle) * size + center.x;
      final y = sin(angle) * size + center.y;

      corners.add(Vector2(x, y));
    }

    return corners;
  }

  @override
  int get hashCode => Object.hash(x, y);

  @override
  bool operator ==(Object other) {
    return (other is Hex) && x == other.x && y == other.y;
  }

  @override
  String toString() {
    return "($x,$y)";
  }

  Map<String, int> toJson() {
    return {"x": x, "y": y};
  }
}
